File: nodejs/tests/capabilities.test.ts
import { describe, it, expect } from 'vitest';
import { join, dirname } from 'node:path';
import { fileURLToPath } from 'node:url';
import { mkdtempSync, rmSync } from 'node:fs';
import { tmpdir } from 'node:os';
import {
  openMachineConfig,
  createMachineConfig,
  supportedFileVersions,
  isOk,
  ok,
  SetMode,
  CREATE,
  type MachineConfigFile,
} from '../src/capabilities/index.js';
import { MachineConfigReader } from '../src/reader.js';
import { MachineConfigWriter } from '../src/writer.js';

const __dirname = dirname(fileURLToPath(import.meta.url));
const FIXTURE = join(__dirname, '..', '..', 'fixtures', 'reference_config.h5');
const FIXTURE_OPCUA = join(
  __dirname,
  '..',
  '..',
  'fixtures',
  'reference_config_opcua.h5',
);
const FIXTURE_OPCUA_MISSING_REQUIRED = join(
  __dirname,
  '..',
  '..',
  'docs',
  'validation',
  'fixtures',
  'opcua_missing_required.h5',
);

describe('capability facade (File_Version 1.0)', () => {
  it('lists supported versions', () => {
    expect(supportedFileVersions()).toContain('1.0');
  });

  it('createMachineConfig rejects an unregistered version', () => {
    const result = createMachineConfig('2.0');
    expect(result.ok).toBe(false);
    if (!result.ok) expect(result.error.code).toBe('UnsupportedVersion');
  });

  it('createMachineConfig dispatches via a registered version', () => {
    // Proves CREATE is genuinely consulted (DISPATCH_REGISTRY_PLAN.md), not
    // just a hardcoded "1.0" check that happens to still work.
    const sentinel = { marker: 'sentinel' } as unknown as MachineConfigFile;
    CREATE['9.9-test'] = () => ok(sentinel);
    try {
      const result = createMachineConfig('9.9-test');
      expect(result.ok).toBe(true);
      if (result.ok) expect(result.value).toBe(sentinel);
    } finally {
      delete CREATE['9.9-test'];
    }
  });

  it('open → getScanner fields match reader JSON', async () => {
    const opened = await openMachineConfig(FIXTURE);
    expect(opened.ok).toBe(true);
    if (!opened.ok) return;
    const file = opened.value;
    expect(file.fileVersion()).toBe('1.0');

    const reader = new MachineConfigReader(FIXTURE);
    const json = await reader.parse();
    const train = file.opticalTrain(0);
    expect(train.ok).toBe(true);
    if (!train.ok) return;
    const scanner = train.value.getScanner();
    expect(scanner.working_distance).toBe(json.optical_trains[0].scanner.working_distance);
    expect(scanner.manufacturer).toBe(json.optical_trains[0].scanner.manufacturer);
    file.close();
  });

  it('Merge setScanner preserves untouched fields after save/reopen', async () => {
    const dir = mkdtempSync(join(tmpdir(), 'mcl-cap-'));
    const out = join(dir, 'merge.h5');
    try {
      const opened = await openMachineConfig(FIXTURE);
      expect(opened.ok).toBe(true);
      if (!opened.ok) return;
      const file = opened.value;
      const train = file.opticalTrain(0);
      expect(train.ok).toBe(true);
      if (!train.ok) return;
      const before = train.value.getScanner();
      const manufacturer = before.manufacturer;
      await train.value.setScanner(
        { ...before, working_distance: 123.5 },
        SetMode.Merge,
      );
      expect(isOk(await file.save(out))).toBe(true);
      file.close();

      const again = await openMachineConfig(out);
      expect(again.ok).toBe(true);
      if (!again.ok) return;
      const sc = again.value.opticalTrain(0);
      expect(sc.ok).toBe(true);
      if (!sc.ok) return;
      const after = sc.value.getScanner();
      expect(after.working_distance).toBe(123.5);
      expect(after.manufacturer).toBe(manufacturer);
      again.value.close();
    } finally {
      rmSync(dir, { recursive: true, force: true });
    }
  });

  it('Replace setScanner replaces whole node including extra', async () => {
    const opened = await openMachineConfig(FIXTURE);
    expect(opened.ok).toBe(true);
    if (!opened.ok) return;
    const train = opened.value.opticalTrain(0);
    expect(train.ok).toBe(true);
    if (!train.ok) return;
    const before = train.value.getScanner();
    const replacement = {
      manufacturer: 'ReplaceCo',
      model: 'ReplaceModel',
      serial_number: 'R-1',
      working_distance: 1,
      working_distance_unit: 'mm',
      scan_field_x: 1,
      scan_field_x_unit: 'mm',
      scan_field_y: 1,
      scan_field_y_unit: 'mm',
      scan_field_z: 1,
      scan_field_z_unit: 'mm',
      scan_head_offset_x: 0,
      scan_head_offset_x_unit: 'mm',
      scan_head_offset_y: 0,
      scan_head_offset_y_unit: 'mm',
      scan_head_offset_z: 0,
      scan_head_offset_z_unit: 'mm',
      extra: {},
    };
    const set = train.value.setScanner(replacement as typeof before, SetMode.Replace);
    expect(set.ok).toBe(true);
    const after = train.value.getScanner();
    expect(after.manufacturer).toBe('ReplaceCo');
    expect(after.working_distance).toBe(1);
    expect(after.extra ?? {}).toEqual({});
    expect(after.model).not.toBe(before.model);
    opened.value.close();
  });

  it('opticalTrains length and InvalidIndex', async () => {
    const opened = await openMachineConfig(FIXTURE);
    expect(opened.ok).toBe(true);
    if (!opened.ok) return;
    expect(opened.value.opticalTrains().length).toBeGreaterThanOrEqual(1);
    const bad = opened.value.opticalTrain(999);
    expect(bad.ok).toBe(false);
    if (!bad.ok) expect(bad.error.code).toBe('InvalidIndex');
    opened.value.close();
  });

  it('opcua NotPresent vs present', async () => {
    const noOpc = await openMachineConfig(FIXTURE);
    expect(noOpc.ok).toBe(true);
    if (!noOpc.ok) return;
    const missing = noOpc.value.opcua();
    expect(missing.ok).toBe(false);
    if (!missing.ok) expect(missing.error.code).toBe('NotPresent');
    noOpc.value.close();

    const withOpc = await openMachineConfig(FIXTURE_OPCUA);
    expect(withOpc.ok).toBe(true);
    if (!withOpc.ok) return;
    const present = withOpc.value.opcua();
    expect(present.ok).toBe(true);
    if (present.ok) {
      const model = present.value.getModel();
      expect(model).toBeTruthy();
    }
    withOpc.value.close();
  });

  it('opcua() Ok when all required fields present on reference_opcua fixture', async () => {
    const opened = await openMachineConfig(FIXTURE_OPCUA);
    expect(opened.ok).toBe(true);
    if (!opened.ok) return;
    const result = opened.value.opcua();
    expect(result.ok).toBe(true);
    if (result.ok) {
      const model = result.value.getModel();
      expect(model.client.machine_profile).not.toBeNull();
      expect(Object.values(model.triggers).every((t) => t.event != null)).toBe(true);
    }
    opened.value.close();
  });

  it('opcua() reports all seven missing required fields at once', async () => {
    const opened = await openMachineConfig(FIXTURE_OPCUA_MISSING_REQUIRED);
    expect(opened.ok).toBe(true);
    if (!opened.ok) return;
    const result = opened.value.opcua();
    expect(result.ok).toBe(false);
    if (result.ok) return;
    expect(result.error.code).toBe('ValidationError');

    const expected = new Set([
      'Machine_Profile',
      'Root_Node',
      'Configure_Client',
      'Pipe_Name',
      'Triggers_Enabled',
      'Trigger_Stop_Ceiling_Layers',
      'Laser Emission Interlock.Event',
    ]);
    expect(new Set(result.error.details)).toEqual(expected);
    expect(result.error.details?.some((d) => d.startsWith('Chamber Oxygen Level'))).toBe(false);
    opened.value.close();
  });

  it('opcua() never reports an optional field, even when genuinely absent', async () => {
    // opcua_missing_required.h5 only clears the 7 required fields — every
    // optional field is still present there, so absence of an optional field
    // from `details` would be trivially true. Also clear an optional field
    // (keep_alive_count) in memory, re-write to a temp file, and confirm
    // `details` still names exactly the same 7 items, not 8.
    const dir = mkdtempSync(join(tmpdir(), 'mcl-cap-opcua-'));
    const out = join(dir, 'missing_required_plus_optional.h5');
    try {
      const reader = new MachineConfigReader(FIXTURE_OPCUA_MISSING_REQUIRED);
      const config = await reader.parse();
      config.opcua!.client.keep_alive_count = null;
      await new MachineConfigWriter(config).write(out);

      const opened = await openMachineConfig(out);
      expect(opened.ok).toBe(true);
      if (!opened.ok) return;
      const result = opened.value.opcua();
      expect(result.ok).toBe(false);
      if (result.ok) return;
      expect(result.error.details?.some((d) => d.includes('Keep_Alive_Count'))).toBe(false);
      expect(result.error.details).toHaveLength(7);
      opened.value.close();
    } finally {
      rmSync(dir, { recursive: true, force: true });
    }
  });

  it('optionalComponents null vs clearbox present', async () => {
    const opened = await openMachineConfig(FIXTURE);
    expect(opened.ok).toBe(true);
    if (!opened.ok) return;
    const train = opened.value.opticalTrain(0);
    expect(train.ok).toBe(true);
    if (!train.ok) return;
    const oc = train.value.optionalComponents();
    expect(oc).not.toBeNull();
    if (!oc) return;
    const cb = oc.clearbox();
    expect(cb.ok).toBe(true);
    opened.value.close();

    const created = createMachineConfig('1.0');
    expect(created.ok).toBe(true);
    if (!created.ok) return;
    // Mock builder defaults include clearbox; strip it to assert null path.
    const t0 = created.value.opticalTrain(0);
    expect(t0.ok).toBe(true);
    if (!t0.ok) return;
    const model = t0.value.getModel();
    (model as { optional_components: { clearbox: null } }).optional_components = {
      clearbox: null,
    };
    t0.value.setModel(model, SetMode.Replace);
    expect(t0.value.optionalComponents()).toBeNull();
    created.value.close();
  });

  it('getCorrectionData/getInverseCorrectionData match Reader', async () => {
    const opened = await openMachineConfig(FIXTURE);
    expect(opened.ok).toBe(true);
    if (!opened.ok) return;
    const file = opened.value;

    const reader = new MachineConfigReader(FIXTURE);
    const expected = await reader.getCorrectionData(0);
    const result = file.getCorrectionData(0);
    expect(result.ok).toBe(true);
    if (!result.ok) return;
    expect(result.value.shape).toEqual(expected.shape);
    expect(Array.from(result.value.data)).toEqual(Array.from(expected.data));

    const expectedInv = await reader.getInverseCorrectionData(0);
    const resultInv = file.getInverseCorrectionData(0);
    expect(resultInv.ok).toBe(true);
    if (!resultInv.ok) return;
    expect(resultInv.value.shape).toEqual(expectedInv.shape);
    expect(Array.from(resultInv.value.data)).toEqual(Array.from(expectedInv.data));
    file.close();
  });

  it('getCorrectionData shape and NaN present through facade', async () => {
    const opened = await openMachineConfig(FIXTURE);
    expect(opened.ok).toBe(true);
    if (!opened.ok) return;
    const grid = opened.value.getCorrectionData(0);
    expect(grid.ok).toBe(true);
    if (!grid.ok) return;
    expect(grid.value.shape).toEqual([257, 257, 2]);
    expect(Array.from(grid.value.data).some((v) => Number.isNaN(v))).toBe(true);
    opened.value.close();
  });

  it('getCorrectionData missing clearbox is NotPresent', async () => {
    const dir = mkdtempSync(join(tmpdir(), 'mcl-cap-cd-'));
    const out = join(dir, 'no_clearbox.h5');
    try {
      // Every stock fixture's trains have a ClearBox, so build one without:
      // take the reference config, strip train 0's ClearBox, re-write.
      const reader = new MachineConfigReader(FIXTURE);
      const config = await reader.parse();
      config.optical_trains[0].optional_components.clearbox = null;
      await new MachineConfigWriter(config).write(out);

      const opened = await openMachineConfig(out);
      expect(opened.ok).toBe(true);
      if (!opened.ok) return;
      const result = opened.value.getCorrectionData(0);
      expect(result.ok).toBe(false);
      if (!result.ok) expect(result.error.code).toBe('NotPresent');
      const invResult = opened.value.getInverseCorrectionData(0);
      expect(invResult.ok).toBe(false);
      if (!invResult.ok) expect(invResult.error.code).toBe('NotPresent');
      opened.value.close();
    } finally {
      rmSync(dir, { recursive: true, force: true });
    }
  });

  it('getCorrectionData works on create()-based instance without touching disk', () => {
    // The specific case that rules out delegate-to-Reader-by-reopening: a
    // create()-d facade has no path at all, so this must convert the
    // already-loaded in-memory model, not re-read from anywhere.
    const created = createMachineConfig('1.0');
    expect(created.ok).toBe(true);
    if (!created.ok) return;
    const file = created.value;
    const grid = file.getCorrectionData(0);
    expect(grid.ok).toBe(true);
    if (grid.ok) expect(grid.value.shape).toEqual([257, 257, 2]);
    const invGrid = file.getInverseCorrectionData(0);
    expect(invGrid.ok).toBe(true);
    if (invGrid.ok) expect(invGrid.value.shape).toEqual([257, 257, 2]);
    file.close();
  });

  it('create(1.0) → set machine name → save → reopen keeps version', async () => {
    const dir = mkdtempSync(join(tmpdir(), 'mcl-cap-'));
    const out = join(dir, 'created.h5');
    try {
      const created = createMachineConfig('1.0');
      expect(created.ok).toBe(true);
      if (!created.ok) return;
      const file = created.value;
      expect(file.fileVersion()).toBe('1.0');
      const meta = file.meta().getModel();
      file.meta().setModel({ ...meta, machine_name: 'CreatedMachine' }, SetMode.Merge);
      expect(isOk(await file.save(out))).toBe(true);
      file.close();

      const again = await openMachineConfig(out);
      expect(again.ok).toBe(true);
      if (!again.ok) return;
      expect(again.value.fileVersion()).toBe('1.0');
      expect(again.value.meta().getModel().machine_name).toBe('CreatedMachine');
      again.value.close();
    } finally {
      rmSync(dir, { recursive: true, force: true });
    }
  });
});

File: rust/tests/capabilities_test.rs
//! Stable model facade tests (File_Version 1.0).

use machine_config::capabilities::{
    create_machine_config, open_machine_config, supported_file_versions, SetMode,
};
use machine_config::capabilities::errors::CapabilityError;
use machine_config::reader::MachineConfigReader;
use machine_config::writer::MachineConfigWriter;
use tempfile::NamedTempFile;

static REFERENCE: &str = concat!(
    env!("CARGO_MANIFEST_DIR"),
    "/../fixtures/reference_config.h5"
);
static REFERENCE_OPCUA: &str = concat!(
    env!("CARGO_MANIFEST_DIR"),
    "/../fixtures/reference_config_opcua.h5"
);
static OPCUA_MISSING_REQUIRED: &str = concat!(
    env!("CARGO_MANIFEST_DIR"),
    "/../docs/validation/fixtures/opcua_missing_required.h5"
);
static REFERENCE_SENSORS: &str = concat!(
    env!("CARGO_MANIFEST_DIR"),
    "/../fixtures/reference_config_synchronous_sensors.h5"
);
static REFERENCE_OPCUA_SENSORS: &str = concat!(
    env!("CARGO_MANIFEST_DIR"),
    "/../fixtures/reference_config_opcua_synchronous_sensors.h5"
);

#[test]
fn supported_versions() {
    assert!(supported_file_versions().contains(&"1.0"));
}

#[test]
fn open_get_scanner_matches_reader() {
    let file = open_machine_config(REFERENCE).unwrap();
    assert_eq!(file.file_version(), "1.0");
    let json = MachineConfigReader::open(REFERENCE).unwrap().parse().unwrap();
    let scanner = file.get_scanner(0).unwrap();
    assert_eq!(
        scanner.working_distance,
        json.optical_trains[0].scanner.working_distance
    );
    assert_eq!(
        scanner.manufacturer,
        json.optical_trains[0].scanner.manufacturer
    );
}

#[test]
fn merge_set_scanner_roundtrip() {
    let mut file = open_machine_config(REFERENCE).unwrap();
    let mut scanner = file.get_scanner(0).unwrap();
    let manufacturer = scanner.manufacturer.clone();
    scanner.working_distance = Some(123.5);
    file.set_scanner(0, scanner, SetMode::Merge).unwrap();
    let tmp = NamedTempFile::new().unwrap();
    file.save(Some(tmp.path())).unwrap();

    let again = open_machine_config(tmp.path()).unwrap();
    let after = again.get_scanner(0).unwrap();
    assert_eq!(after.working_distance, Some(123.5));
    assert_eq!(after.manufacturer, manufacturer);
}

#[test]
fn replace_set_scanner() {
    let mut file = open_machine_config(REFERENCE).unwrap();
    let before = file.get_scanner(0).unwrap();
    assert!(before.scan_field_x.is_some());
    let mut replacement = before.clone();
    replacement.manufacturer = "ReplaceCo".into();
    replacement.working_distance = Some(1.0);
    replacement.scan_field_x = None;
    file.set_scanner(0, replacement, SetMode::Replace).unwrap();
    let after = file.get_scanner(0).unwrap();
    assert_eq!(after.manufacturer, "ReplaceCo");
    assert_eq!(after.working_distance, Some(1.0));
    assert!(after.scan_field_x.is_none());
}

#[test]
fn invalid_index() {
    let file = open_machine_config(REFERENCE).unwrap();
    let err = file.get_train(999).unwrap_err();
    assert!(matches!(err, CapabilityError::InvalidIndex(_)));
}

#[test]
fn opcua_not_present_vs_present() {
    let no_opc = open_machine_config(REFERENCE).unwrap();
    assert!(matches!(
        no_opc.get_opcua().unwrap_err(),
        CapabilityError::NotPresent(_)
    ));

    let with_opc = open_machine_config(REFERENCE_OPCUA).unwrap();
    assert!(with_opc.get_opcua().is_ok());
}

#[test]
fn opcua_required_fields_present_on_reference_fixture() {
    let file = open_machine_config(REFERENCE_OPCUA).unwrap();
    let opcua = file.get_opcua().expect("all required fields are present on this fixture");
    // Sanity-check a couple of the fields the check depends on, so a future
    // accidental fixture edit fails loudly here rather than only downstream.
    assert!(opcua.client.machine_profile.is_some());
    assert!(opcua.triggers.values().all(|t| t.event.is_some()));
}

#[test]
fn opcua_missing_required_fields_reports_all_seven_at_once() {
    let file = open_machine_config(OPCUA_MISSING_REQUIRED).unwrap();
    let err = file.get_opcua().unwrap_err();
    let details = match err {
        CapabilityError::ValidationError { details: Some(details), .. } => details,
        other => panic!("expected ValidationError with details, got {other:?}"),
    };

    let mut actual = details.clone();
    actual.sort();
    let mut expected = vec![
        "Machine_Profile".to_string(),
        "Root_Node".to_string(),
        "Configure_Client".to_string(),
        "Pipe_Name".to_string(),
        "Triggers_Enabled".to_string(),
        "Trigger_Stop_Ceiling_Layers".to_string(),
        "Laser Emission Interlock.Event".to_string(),
    ];
    expected.sort();
    assert_eq!(actual, expected, "details must name exactly the seven missing fields");

    // The other trigger still has Event — must not be reported as missing.
    assert!(!details.iter().any(|d| d.starts_with("Chamber Oxygen Level")));
}

#[test]
fn opcua_optional_field_never_appears_in_missing_details() {
    // opcua_missing_required.h5 only clears the 7 required fields — every
    // optional field is still present there, so absence of an optional field
    // from `details` would be trivially true. To make this a real check
    // (not a tautology), also clear an optional field (keep_alive_count) in
    // memory, re-write to a temp file, and confirm `details` still names
    // exactly the same 7 required fields — not 8.
    let mut config = MachineConfigReader::open(OPCUA_MISSING_REQUIRED)
        .unwrap()
        .parse()
        .unwrap();
    config.opcua.as_mut().unwrap().client.keep_alive_count = None;
    let tmp = NamedTempFile::with_suffix(".h5").unwrap();
    MachineConfigWriter::new(&config).write(tmp.path()).unwrap();

    let file = open_machine_config(tmp.path()).unwrap();
    let err = file.get_opcua().unwrap_err();
    let details = match err {
        CapabilityError::ValidationError { details: Some(details), .. } => details,
        other => panic!("expected ValidationError with details, got {other:?}"),
    };

    assert!(
        !details.iter().any(|d| d.contains("Keep_Alive_Count")),
        "optional field must never appear in details, even when genuinely absent: {details:?}"
    );
    assert_eq!(details.len(), 7, "clearing an optional field must not change the missing count");
}

#[test]
fn clearbox_optional() {
    let file = open_machine_config(REFERENCE).unwrap();
    assert!(file.get_clearbox(0).unwrap().is_some());
}

#[test]
fn clearbox_synchronous_sensors_empty_when_fixture_has_none() {
    // reference_config.h5 deliberately has no Synchronous_Sensors group at
    // all (see SYNCHRONOUS_SENSOR_PLAN.md Phase 0) — the map must read back
    // empty, not missing/absent, since there is no separate "absent" state.
    let file = open_machine_config(REFERENCE).unwrap();
    let cb = file.get_clearbox(0).unwrap().unwrap();
    assert!(cb.synchronous_sensors.is_empty());
}

#[test]
fn clearbox_synchronous_sensors_present_and_named_on_dedicated_fixture() {
    let file = open_machine_config(REFERENCE_SENSORS).unwrap();
    let cb = file.get_clearbox(0).unwrap().unwrap();
    assert_eq!(cb.synchronous_sensors.len(), 1);
    let sensor = cb.synchronous_sensors.get("Oxygen Sensor")
        .expect("key is this fixture's arbitrary label, not a schema-significant name");
    assert_eq!(sensor.sensor_name, Some("ZR800 Oxygen Analyzer".to_string()));
    assert_eq!(sensor.port_id, Some(5));
}

#[test]
fn combined_fixture_has_opcua_and_synchronous_sensors_through_facade() {
    // Third-layer check for the combined fixture: hdf5.rs's own test module
    // proves the internal adapter, reader.rs proves the plain public
    // MachineConfigReader, this proves the capabilities facade too.
    let file = open_machine_config(REFERENCE_OPCUA_SENSORS).unwrap();
    assert!(file.get_opcua().is_ok(), "OPCUA must be present on the combined fixture");
    let cb = file.get_clearbox(0).unwrap().unwrap();
    assert_eq!(
        cb.synchronous_sensors["Oxygen Sensor"].sensor_name,
        Some("ZR800 Oxygen Analyzer".to_string())
    );
}

/// `CorrectionData` derives `PartialEq`, but IEEE 754 `NaN != NaN`, so a
/// direct `assert_eq!` on real grid data (which always contains NaN cells)
/// fails even when the two buffers are bit-for-bit identical. Compare shape
/// plus element-wise, treating "both NaN" as equal.
fn assert_correction_data_eq(
    a: &machine_config::models::CorrectionData,
    b: &machine_config::models::CorrectionData,
) {
    assert_eq!(a.shape, b.shape);
    assert_eq!(a.data.len(), b.data.len());
    for (x, y) in a.data.iter().zip(b.data.iter()) {
        if x.is_nan() || y.is_nan() {
            assert!(x.is_nan() && y.is_nan(), "NaN mismatch: {x} vs {y}");
        } else {
            assert_eq!(x, y);
        }
    }
}

#[test]
fn get_correction_data_matches_reader() {
    let file = open_machine_config(REFERENCE).unwrap();
    let reader = MachineConfigReader::open(REFERENCE).unwrap();
    assert_correction_data_eq(
        &file.get_correction_data(0).unwrap(),
        &reader.get_correction_data(0).unwrap(),
    );
    assert_correction_data_eq(
        &file.get_inverse_correction_data(0).unwrap(),
        &reader.get_inverse_correction_data(0).unwrap(),
    );
}

#[test]
fn get_correction_data_shape_and_nan_present_through_facade() {
    let file = open_machine_config(REFERENCE).unwrap();
    let cd = file.get_correction_data(0).unwrap();
    assert_eq!(cd.shape, [257, 257, 2]);
    assert!(cd.data.iter().any(|v| v.is_nan()));
}

#[test]
fn get_correction_data_missing_clearbox_is_not_present() {
    // Every stock fixture's trains have a ClearBox, so build one without: take
    // the reference config, strip train 0's ClearBox, re-write to a temp file.
    let mut config = MachineConfigReader::open(REFERENCE).unwrap().parse().unwrap();
    config.optical_trains[0].optional_components.clearbox = None;
    let tmp = NamedTempFile::with_suffix(".h5").unwrap();
    MachineConfigWriter::new(&config).write(tmp.path()).unwrap();

    let file = open_machine_config(tmp.path()).unwrap();
    assert!(file.get_clearbox(0).unwrap().is_none());
    assert!(matches!(
        file.get_correction_data(0).unwrap_err(),
        CapabilityError::NotPresent(_)
    ));
    assert!(matches!(
        file.get_inverse_correction_data(0).unwrap_err(),
        CapabilityError::NotPresent(_)
    ));
}

#[test]
fn get_correction_data_works_on_create_based_instance_without_touching_disk() {
    // The specific case that rules out delegate-to-Reader-by-reopening: a
    // create()-d facade has no path at all, so this must convert the
    // already-loaded in-memory model, not re-read from anywhere.
    let file = create_machine_config("1.0").unwrap();
    let cd = file.get_correction_data(0).unwrap();
    assert_eq!(cd.shape, [257, 257, 2]);
    let icd = file.get_inverse_correction_data(0).unwrap();
    assert_eq!(icd.shape, [257, 257, 2]);
}

#[test]
fn create_set_meta_save_reopen() {
    let mut file = create_machine_config("1.0").unwrap();
    assert_eq!(file.file_version(), "1.0");
    let mut meta = file.get_meta().unwrap();
    meta.machine_name = "CreatedMachine".into();
    file.set_meta(meta, SetMode::Merge).unwrap();
    let tmp = NamedTempFile::new().unwrap();
    file.save(Some(tmp.path())).unwrap();

    let again = open_machine_config(tmp.path()).unwrap();
    assert_eq!(again.file_version(), "1.0");
    assert_eq!(again.get_meta().unwrap().machine_name, "CreatedMachine");
}
